blockchain/message_pool/src/msgpool.rs

```diff
@@ -0,0 +1,1033 @@
+            }
+        }
+
+        self.local_msgs = msg_vec;
+
+        Ok(())
+    }
+}
+
+/// Remove a message from pending given the from address and sequence
+pub async fn remove(
+    from: &Address,
+    pending: Arc<RwLock<HashMap<Address, MsgSet>>>,
+    sequence: u64,
+) -> Result<(), Error> {
+    let mut pending = pending.write().await;
+
+    let mset = pending
+        .get_mut(from)
+        .ok_or_else(|| Error::InvalidFromAddr)?;
+    mset.msgs.remove(&sequence);
+
+    if mset.msgs.is_empty() {
+        pending.remove(from);
+    } else {
+        let mut max_sequence: u64 = 0;
+        for sequence_temp in mset.msgs.keys().cloned() {
+            if max_sequence < sequence_temp {
+                max_sequence = sequence_temp;
+            }
+        }
+        if max_sequence < sequence {
+            max_sequence = sequence;
+        }
+        mset.next_sequence = max_sequence + 1;
+    }
+    Ok(())
+}
+
+/// Attempt to get a signed message that corresponds to an unsigned message in bls_sig_cache
+async fn recover_sig(
+    bls_sig_cache: &mut LruCache<Cid, Signature>,
+    msg: UnsignedMessage,
+) -> Result<SignedMessage, Error> {
+    let val = bls_sig_cache
+        .get(&msg.cid()?)
+        .ok_or_else(|| Error::Other("Could not recover sig".to_owned()))?;
+    Ok(SignedMessage::new_from_parts(msg, val.clone()))
+}
+
+/// Finish verifying signed message before adding it to the pending mset hashmap. If an entry
+/// in the hashmap does not yet exist, create a new mset that will correspond to the from message
+/// and push it to the pending hashmap
+async fn add_helper<T>(
+    api: RwLockReadGuard<'_, T>,
+    mut bls_sig_cache: RwLockWriteGuard<'_, LruCache<Cid, Signature>>,
+    mut pending: RwLockWriteGuard<'_, HashMap<Address, MsgSet>>,
+    msg: SignedMessage,
+) -> Result<(), Error>
+where
+    T: Provider,
+{
+    if msg.signature().signature_type() == SignatureType::BLS {
+        bls_sig_cache.put(msg.cid()?, msg.signature().clone());
+    }
+
+    if msg.message().gas_limit() > 100_000_000 {
+        return Err(Error::Other(
+            "given message has too high of a gas limit".to_string(),
+        ));
+    }
+
+    api.put_message(&msg)?;
+
+    // let mut pending = pending.write().await;
+    let msett = pending.get_mut(msg.message().from());
+    match msett {
+        Some(mset) => mset.add(msg)?,
+        None => {
+            let mut mset = MsgSet::new();
+            let from = *msg.message().from();
+            mset.add(msg)?;
+            pending.insert(from, mset);
+        }
+    }
+
+    Ok(())
+}
+
+/// This function will revert and/or apply tipsets to the message pool. This function should be
+/// called every time that there is a head change in the message pool
+pub async fn head_change<T>(
+    api: Arc<RwLock<T>>,
+    bls_sig_cache: Arc<RwLock<LruCache<Cid, Signature>>>,
+    pending: Arc<RwLock<HashMap<Address, MsgSet>>>,
+    cur_tipset: Arc<RwLock<Tipset>>,
```

Comment:
pretty sure these don't need to be cloned arcs also, and can just be references to the RwLock<..> to avoid the atomic count (minimal change)